src/util/sharing_map.h

```diff
@@ -887,105 +877,46 @@ ::erase_all(const keyst &ks, const tvt &key_exists)
   return cnt;
 }
 
-/// Insert element, return const reference
+/// Insert element, element must not exist in map
 ///
 /// Complexity:
 /// - Worst case: O(H * S + M)
 /// - Best case: O(H)
 ///
 /// \param k: The key of the element to insert
 /// \param m: The mapped value to insert
-/// \param key_exists: Hint to indicate whether the element is known to exist
-///   (possible values `false` or `unknown`)
-/// \return Pair of const reference to existing or newly inserted element, and
-///   boolean indicating if new element was inserted
-SHARING_MAPT2(const, const_find_type)
-::insert(const key_type &k, const mapped_type &m, const tvt &key_exists)
+SHARING_MAPT4(U, void)
+::insert(const key_type &k, U &&m)
 {
-  SM_ASSERT(!key_exists.is_true());
-  SM_ASSERT(!key_exists.is_false() || !has_key(k));
-
-  if(key_exists.is_unknown())
-  {
-    const leaft *lp = as_const(this)->get_leaf_node(k);
-    if(lp != nullptr)
-      return const_find_type(lp->get_value(), false);
-  }
-
   innert *cp = get_container_node(k);
   SM_ASSERT(cp != nullptr);
 
-  const leaft *lp = cp->place_leaf(k, m);
+  cp->place_leaf(k, std::forward<U>(m));
   num++;
-
-  return const_find_type(lp->get_value(), true);
 }
 
-// Insert element, return const reference
-SHARING_MAPT2(const, const_find_type)
-::insert(const value_type &p, const tvt &key_exists)
-{
-  return insert(p.first, p.second, key_exists);
-}
-
-/// Insert element, return non-const reference
+/// Replace element, element must exist in map
 ///
 /// Complexity:
 /// - Worst case: O(H * S + M)
 /// - Best case: O(H)
 ///
 /// \param k: The key of the element to insert
-/// \param m: The mapped value to insert
-/// \return Pair of reference to existing or newly inserted element, and boolean
-///   indicating if new element was inserted
-SHARING_MAPT2(const, find_type)::place(const key_type &k, const mapped_type &m)
+/// \param m: The mapped value to replace the old value with
+SHARING_MAPT4(U, void)
+::replace(const key_type &k, U &&m)
 {
   innert *cp = get_container_node(k);
   SM_ASSERT(cp != nullptr);
 
   leaft *lp = cp->find_leaf(k);
+  PRECONDITION(lp != nullptr); // key must exist in map
 
-  if(lp != nullptr)
-    return find_type(lp->get_value(), false);
-
-  lp = cp->place_leaf(k, m);
-  num++;
+  INVARIANT(
+    value_equalt()(as_const(lp)->get_value(), m),
+    "values should not be replaced with equal values to maximize sharing");
```

Comment:
Wouldn't a slightly less fancy `if(check_equal) ...` also do the trick instead of the `struct True ...` above?

Reply:
This wouldn't compile when `valueT` doesn't have a defined equality operator, as the code in the `if(check_equal) ...` branch is still parsed when `check_equal` is false.

Reply:
Well, note that it does always parse, it just wouldn't type check. And I don't think code that isn't instantiated would be subject to type checking (as in general that wouldn't work).

Reply:
A branch dependent on a boolean template parameter is still typechecked even when statically false. I've had this issue several times before.

Reply:
Ok, thanks!